// blis-models/lib/Action.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionTypes = {
    TEXT: "TEXT",
    API_LOCAL: "API_LOCAL",
    API_AZURE: "API_AZURE",
    INTENT: "INTENT",
    CARD: "CARD"
};
var ActionMetaData = (function () {
    function ActionMetaData(init) {
        Object.assign(this, init);
    }
    ActionMetaData.prototype.Equal = function (metaData) {
        return (this.actionType == metaData.actionType);
    };
    return ActionMetaData;
}());
exports.ActionMetaData = ActionMetaData;
var ActionBase = (function () {
    function ActionBase(init) {
        this.metadata = new ActionMetaData();
        Object.assign(this, init);
    }
    return ActionBase;
}());
exports.ActionBase = ActionBase;
var ActionList = (function () {
    function ActionList(init) {
        Object.assign(this, init);
    }
    return ActionList;
}());
exports.ActionList = ActionList;
var ActionIdList = (function () {
    function ActionIdList(init) {
        Object.assign(this, init);
    }
    return ActionIdList;
}());
exports.ActionIdList = ActionIdList;
