// blis-models/lib/Entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EntityType;
(function (EntityType) {
    EntityType["LOCAL"] = "LOCAL";
    EntityType["LUIS"] = "LUIS";
})(EntityType = exports.EntityType || (exports.EntityType = {}));
;
var EntityMetaData = (function () {
    function EntityMetaData(init) {
        Object.assign(this, init);
    }
    /** Make negate of given metadata */
    EntityMetaData.prototype.MakeNegative = function (posId) {
        return new EntityMetaData({ isBucket: this.isBucket, negativeId: null, positiveId: posId });
    };
    return EntityMetaData;
}());
exports.EntityMetaData = EntityMetaData;
var EntityBase = (function () {
    function EntityBase(init) {
        Object.assign(this, init);
    }
    return EntityBase;
}());
exports.EntityBase = EntityBase;
var EntityList = (function () {
    function EntityList(init) {
        Object.assign(this, init);
    }
    return EntityList;
}());
exports.EntityList = EntityList;
var EntityIdList = (function () {
    function EntityIdList(init) {
        Object.assign(this, init);
    }
    return EntityIdList;
}());
exports.EntityIdList = EntityIdList;
