// blis-models/lib/LogDialog.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Extract_1 = require("./Extract");
var LogExtractorStep = (function (_super) {
    __extends(LogExtractorStep, _super);
    function LogExtractorStep(init) {
        var _this = _super.call(this, init) || this;
        Object.assign(_this, init);
        return _this;
    }
    return LogExtractorStep;
}(Extract_1.ExtractResponse));
exports.LogExtractorStep = LogExtractorStep;
var LogScorerStep = (function () {
    function LogScorerStep(init) {
        Object.assign(this, init);
    }
    return LogScorerStep;
}());
exports.LogScorerStep = LogScorerStep;
var LogRound = (function () {
    function LogRound(init) {
        Object.assign(this, init);
    }
    return LogRound;
}());
exports.LogRound = LogRound;
var LogDialog = (function () {
    function LogDialog(init) {
        Object.assign(this, init);
    }
    return LogDialog;
}());
exports.LogDialog = LogDialog;
var LogDialogList = (function () {
    function LogDialogList(init) {
        Object.assign(this, init);
    }
    return LogDialogList;
}());
exports.LogDialogList = LogDialogList;
var LogDialogIdList = (function () {
    function LogDialogIdList(init) {
        Object.assign(this, init);
    }
    return LogDialogIdList;
}());
exports.LogDialogIdList = LogDialogIdList;
