// blis-models/lib/TrainDialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var LabeledEntity = (function () {
    function LabeledEntity(init) {
        Object.assign(this, init);
    }
    return LabeledEntity;
}());
exports.LabeledEntity = LabeledEntity;
var TextVariation = (function () {
    function TextVariation(init) {
        Object.assign(this, init);
    }
    return TextVariation;
}());
exports.TextVariation = TextVariation;
var TrainExtractorStep = (function () {
    function TrainExtractorStep(init) {
        Object.assign(this, init);
    }
    return TrainExtractorStep;
}());
exports.TrainExtractorStep = TrainExtractorStep;
var TrainScorerStep = (function () {
    function TrainScorerStep(init) {
        Object.assign(this, init);
    }
    return TrainScorerStep;
}());
exports.TrainScorerStep = TrainScorerStep;
var TrainRound = (function () {
    function TrainRound(init) {
        Object.assign(this, init);
    }
    return TrainRound;
}());
exports.TrainRound = TrainRound;
var TrainDialog = (function () {
    function TrainDialog(init) {
        Object.assign(this, init);
    }
    return TrainDialog;
}());
exports.TrainDialog = TrainDialog;
var TrainResponse = (function () {
    function TrainResponse(init) {
        Object.assign(this, init);
    }
    return TrainResponse;
}());
exports.TrainResponse = TrainResponse;
var TrainDialogList = (function () {
    function TrainDialogList(init) {
        Object.assign(this, init);
    }
    return TrainDialogList;
}());
exports.TrainDialogList = TrainDialogList;
var TrainDialogIdList = (function () {
    function TrainDialogIdList(init) {
        Object.assign(this, init);
    }
    return TrainDialogIdList;
}());
exports.TrainDialogIdList = TrainDialogIdList;
