// blis-models/lib/UI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ScoreReason;
(function (ScoreReason) {
    // Action has been masked
    ScoreReason["NotAvailable"] = "notAvailable";
    // Action can't be scored because it hasn't been trained yet
    ScoreReason["NotScorable"] = "notScorable";
    // Score has not yet been calculated
    ScoreReason["NotCalculated"] = "notCalculated";
})(ScoreReason = exports.ScoreReason || (exports.ScoreReason = {}));
var UIScoreInput = (function () {
    function UIScoreInput(init) {
        Object.assign(this, init);
    }
    return UIScoreInput;
}());
exports.UIScoreInput = UIScoreInput;
var UIExtractResponse = (function () {
    function UIExtractResponse(init) {
        Object.assign(this, init);
    }
    return UIExtractResponse;
}());
exports.UIExtractResponse = UIExtractResponse;
var UITeachResponse = (function () {
    function UITeachResponse(init) {
        Object.assign(this, init);
    }
    return UITeachResponse;
}());
exports.UITeachResponse = UITeachResponse;
var UIScoreResponse = (function () {
    function UIScoreResponse(init) {
        Object.assign(this, init);
    }
    return UIScoreResponse;
}());
exports.UIScoreResponse = UIScoreResponse;
var UITrainScorerStep = (function () {
    function UITrainScorerStep(init) {
        Object.assign(this, init);
    }
    return UITrainScorerStep;
}());
exports.UITrainScorerStep = UITrainScorerStep;
