// blisdk/lib/BlisDialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var builder = require("botbuilder");
var blis_models_1 = require("blis-models");
var BlisRecognizer_1 = require("./BlisRecognizer");
var BlisDebug_1 = require("./BlisDebug");
var Utils_1 = require("./Utils");
var BlisMemory_1 = require("./BlisMemory");
var BlisClient_1 = require("./BlisClient");
var BlisContext_1 = require("./BlisContext");
var ClientMemoryManager_1 = require("./Memory/ClientMemoryManager");
var Server_1 = require("./Http/Server");
var AzureFunctions_1 = require("./AzureFunctions");
var DOLRunner_1 = require("./DOLRunner");
exports.BLIS_INTENT_WRAPPER = "BLIS_INTENT_WRAPPER";
var BlisDialog = (function (_super) {
    tslib_1.__extends(BlisDialog, _super);
    function BlisDialog(bot, options) {
        var _this = _super.call(this) || this;
        _this.bot = bot;
        try {
            BlisDebug_1.BlisDebug.InitLogger(bot);
            _this.options = options;
            _this.blisRecognizer = new BlisRecognizer_1.BlisRecognizer();
            _this.recognizers = new builder.IntentRecognizerSet({ recognizers: [_this.blisRecognizer] });
            BlisDebug_1.BlisDebug.Log("Creating client....");
            BlisClient_1.BlisClient.SetServiceURI(options.serviceUri);
            BlisClient_1.BlisClient.Init(options.user, options.secret, options.azureFunctionsUrl, options.azureFunctionsKey);
            BlisMemory_1.BlisMemory.Init(options.redisServer, options.redisKey);
            // If app not set, assume running on localhost init DOL Runner
            if (_this.options.localhost) {
                DOLRunner_1.InitDOLRunner();
            }
            Server_1.Server.Init();
        }
        catch (error) {
            BlisDebug_1.BlisDebug.Error(error);
        }
        return _this;
    }
    // Create singleton
    BlisDialog.Init = function (bot, options) {
        BlisDialog.dialog = new BlisDialog(bot, options);
        return BlisDialog.dialog;
    };
    Object.defineProperty(BlisDialog, "Instance", {
        get: function () {
            return this.dialog;
        },
        enumerable: true,
        configurable: true
    });
    /** Called when a new reply message has been received from a user. */
    BlisDialog.prototype.replyReceived = function (session, recognizeResult) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            var context;
            return tslib_1.__generator(this, function (_a) {
                if (!recognizeResult) {
                    context = session.toRecognizeContext();
                    context.dialogData = session.dialogData;
                    context.activeDialog = true;
                    this.recognize(context, function (error, result) {
                        var blisResult = result;
                        try {
                            if (!error) {
                                _this.invokeAnswer(session, blisResult);
                            }
                        }
                        catch (e) {
                            _this.emitError(session, e);
                        }
                    });
                }
                else {
                    this.invokeAnswer(session, recognizeResult);
                }
                return [2 /*return*/];
            });
        });
    };
    /** Parses the users utterance and assigns a score from 0.0 - 1.0 indicating
     * how confident the dialog is that it understood the users utterance.  */
    BlisDialog.prototype.recognize = function (context, cb) {
        this.recognizers.recognize(context, cb);
    };
    BlisDialog.prototype.recognizer = function (plugin) {
        // Append recognizer
        this.recognizers.recognizer(plugin);
        return this;
    };
    BlisDialog.prototype.invokeAnswer = function (session, recognizeResult) {
        var _this = this;
        this.ProcessInput(session, function () { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/];
            });
        }); });
    };
    BlisDialog.prototype.validationError = function () {
        var errMsg = "";
        if (!this.options.redisKey) {
            errMsg += "Options missing redisKey. Set BLIS_REDIS_KEY Env value.\n\n";
        }
        if (!this.options.redisServer) {
            errMsg += "Options missing redisServer. Set BLIS_REDIS_SERVER Env value.\n\n";
        }
        if (!this.options.localhost && !this.options.appId) {
            errMsg += "Options must specify appId when not running on localhost. Set BLIS_APP_ID Env value.\n\n";
        }
        return errMsg;
        /*
                    serviceUri: serviceUrl,
            user: config.BLIS_USER,
            secret: config.BLIS_SECRET,
            azureFunctionsUrl : config.BLIS_FUNCTIONS_URL*/
    };
    BlisDialog.prototype.ProcessInput = function (session, cb) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var context, memory, validationError, inTeach, app, sessionResponse, error_1, msg, sessionId, userInput, extractResponse, error_2, msg, error_3, msg;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        console.log("Process Input...");
                        context = new BlisContext_1.BlisContext(this.bot, session);
                        console.log("->mem");
                        memory = context.Memory();
                        validationError = this.validationError();
                        if (!validationError) return [3 /*break*/, 2];
                        BlisDebug_1.BlisDebug.Error(validationError);
                        return [4 /*yield*/, Utils_1.Utils.SendMessage(this.bot, memory, validationError)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 17, , 19]);
                        console.log("->inteach");
                        return [4 /*yield*/, memory.BotState.InTeach()];
                    case 3:
                        inTeach = _a.sent();
                        console.log("->app");
                        return [4 /*yield*/, memory.BotState.App()];
                    case 4:
                        app = _a.sent();
                        if (!!app) return [3 /*break*/, 9];
                        if (!this.options.appId) return [3 /*break*/, 9];
                        _a.label = 5;
                    case 5:
                        _a.trys.push([5, 7, , 9]);
                        console.log("Selecting app: " + this.options.appId);
                        return [4 /*yield*/, BlisClient_1.BlisClient.client.StartSession(this.options.appId)];
                    case 6:
                        sessionResponse = _a.sent();
                        memory.StartSession(sessionResponse.sessionId, false);
                        return [3 /*break*/, 9];
                    case 7:
                        error_1 = _a.sent();
                        console.log("ERROR: Can't start session!");
                        msg = BlisDebug_1.BlisDebug.Error(error_1);
                        return [4 /*yield*/, Utils_1.Utils.SendMessage(this.bot, memory, msg)];
                    case 8:
                        _a.sent();
                        return [2 /*return*/];
                    case 9:
                        console.log("Got input");
                        return [4 /*yield*/, memory.BotState.SessionId()];
                    case 10:
                        sessionId = _a.sent();
                        userInput = new blis_models_1.UserInput({ text: session.message.text });
                        if (!!inTeach) return [3 /*break*/, 16];
                        console.log("Calling entity extractor");
                        _a.label = 11;
                    case 11:
                        _a.trys.push([11, 14, , 16]);
                        return [4 /*yield*/, BlisClient_1.BlisClient.client.SessionExtract(app.appId, sessionId, userInput)];
                    case 12:
                        extractResponse = _a.sent();
                        return [4 /*yield*/, this.ProcessExtraction(app.appId, sessionId, memory, extractResponse.text, extractResponse.predictedEntities, extractResponse.definitions.entities)];
                    case 13:
                        _a.sent();
                        return [3 /*break*/, 16];
                    case 14:
                        error_2 = _a.sent();
                        msg = BlisDebug_1.BlisDebug.Error(error_2);
                        return [4 /*yield*/, Utils_1.Utils.SendMessage(this.bot, memory, msg)];
                    case 15:
                        _a.sent();
                        return [3 /*break*/, 16];
                    case 16: return [3 /*break*/, 19];
                    case 17:
                        error_3 = _a.sent();
                        msg = BlisDebug_1.BlisDebug.Error(error_3);
                        return [4 /*yield*/, Utils_1.Utils.SendMessage(this.bot, memory, msg)];
                    case 18:
                        _a.sent();
                        return [3 /*break*/, 19];
                    case 19: return [2 /*return*/];
                }
            });
        });
    };
    BlisDialog.prototype.ProcessExtraction = function (appId, sessionId, memory, text, predictedEntities, allEntities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var scoreInput, scoreResponse, bestAction;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.CallLuisCallback(text, predictedEntities, memory, allEntities)];
                    case 1:
                        scoreInput = _a.sent();
                        return [4 /*yield*/, BlisClient_1.BlisClient.client.SessionScore(appId, sessionId, scoreInput)];
                    case 2:
                        scoreResponse = _a.sent();
                        bestAction = scoreResponse.scoredActions[0];
                        if (!bestAction) return [3 /*break*/, 4];
                        this.TakeAction(bestAction, memory, allEntities);
                        if (!!bestAction.isTerminal) return [3 /*break*/, 4];
                        return [4 /*yield*/, this.ProcessExtraction(appId, sessionId, memory, "", [], allEntities)];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    BlisDialog.prototype.TakeAction = function (scoredAction, memory, allEntities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var actionType, _a;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        actionType = scoredAction.metadata.actionType;
                        _a = actionType;
                        switch (_a) {
                            case blis_models_1.ActionTypes.TEXT: return [3 /*break*/, 1];
                            case blis_models_1.ActionTypes.CARD: return [3 /*break*/, 3];
                            case blis_models_1.ActionTypes.INTENT: return [3 /*break*/, 5];
                            case blis_models_1.ActionTypes.API_AZURE: return [3 /*break*/, 7];
                            case blis_models_1.ActionTypes.API_LOCAL: return [3 /*break*/, 9];
                        }
                        return [3 /*break*/, 11];
                    case 1: return [4 /*yield*/, this.TakeTextAction(scoredAction, memory, allEntities)];
                    case 2:
                        _b.sent();
                        return [3 /*break*/, 11];
                    case 3: return [4 /*yield*/, this.TakeCardAction(scoredAction, memory, allEntities)];
                    case 4:
                        _b.sent();
                        return [3 /*break*/, 11];
                    case 5: return [4 /*yield*/, this.TakeIntentAction(scoredAction, memory, allEntities)];
                    case 6:
                        _b.sent();
                        return [3 /*break*/, 11];
                    case 7: return [4 /*yield*/, this.TakeAzureAPIAction(scoredAction, memory, allEntities)];
                    case 8:
                        _b.sent();
                        return [3 /*break*/, 11];
                    case 9: return [4 /*yield*/, this.TakeLocalAPIAction(scoredAction, memory, allEntities)];
                    case 10:
                        _b.sent();
                        return [3 /*break*/, 11];
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    BlisDialog.prototype.TakeTextAction = function (scoredAction, memory, allEntities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var outText;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.CallBlisCallback(scoredAction, memory, allEntities)];
                    case 1:
                        outText = _a.sent();
                        return [4 /*yield*/, Utils_1.Utils.SendMessage(this.bot, memory, outText)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    BlisDialog.prototype.TakeCardAction = function (scoredAction, memory, allEntities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/];
            });
        });
    };
    BlisDialog.prototype.TakeLocalAPIAction = function (scoredAction, memory, allEntities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var apiName, args, argArray, _i, args_1, arg, _a, _b, api, msg, memoryManager, output;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!BlisDialog.apiCallbacks) {
                            BlisDebug_1.BlisDebug.Error("No Local APIs defined.");
                            return [2 /*return*/];
                        }
                        apiName = blis_models_1.ModelUtils.GetPrimaryPayload(scoredAction);
                        args = blis_models_1.ModelUtils.GetArguments(scoredAction);
                        argArray = [];
                        _i = 0, args_1 = args;
                        _c.label = 1;
                    case 1:
                        if (!(_i < args_1.length)) return [3 /*break*/, 4];
                        arg = args_1[_i];
                        _b = (_a = argArray).push;
                        return [4 /*yield*/, memory.BotMemory.SubstituteEntities(arg)];
                    case 2:
                        _b.apply(_a, [_c.sent()]);
                        _c.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4:
                        api = BlisDialog.apiCallbacks[apiName];
                        if (!api) {
                            msg = BlisDebug_1.BlisDebug.Error("API \"" + apiName + "\" is undefined");
                            throw new Error(msg);
                        }
                        memoryManager = new ClientMemoryManager_1.ClientMemoryManager(memory, allEntities);
                        return [4 /*yield*/, api(memoryManager, argArray)];
                    case 5:
                        output = _c.sent();
                        if (!output) return [3 /*break*/, 7];
                        return [4 /*yield*/, Utils_1.Utils.SendMessage(this.bot, memory, output)];
                    case 6:
                        _c.sent();
                        _c.label = 7;
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    BlisDialog.prototype.TakeAzureAPIAction = function (scoredAction, memory, allEntities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var apiString, funcName, args, entities, output;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        apiString = scoredAction.payload;
                        funcName = apiString.split(' ')[0];
                        args = blis_models_1.ModelUtils.RemoveWords(apiString, 1);
                        return [4 /*yield*/, memory.BotMemory.SubstituteEntities(args)];
                    case 1:
                        entities = _a.sent();
                        return [4 /*yield*/, AzureFunctions_1.AzureFunctions.Call(BlisClient_1.BlisClient.client.azureFunctionsUrl, BlisClient_1.BlisClient.client.azureFunctionsKey, funcName, entities)];
                    case 2:
                        output = _a.sent();
                        if (!output) return [3 /*break*/, 4];
                        return [4 /*yield*/, Utils_1.Utils.SendMessage(this.bot, memory, output)];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    BlisDialog.prototype.TakeIntentAction = function (scoredAction, memory, allEntities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var apiString, intentName, args, entities, session, inTeach;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        apiString = scoredAction.payload;
                        intentName = apiString.split(' ')[0];
                        args = blis_models_1.ModelUtils.RemoveWords(apiString, 1);
                        return [4 /*yield*/, memory.BotMemory.GetEntities(args)];
                    case 1:
                        entities = _a.sent();
                        return [4 /*yield*/, memory.BotState.Session(this.bot)];
                    case 2:
                        session = _a.sent();
                        return [4 /*yield*/, memory.BotState.InTeach()];
                    case 3:
                        inTeach = _a.sent();
                        if (inTeach) {
                            session.beginDialog(exports.BLIS_INTENT_WRAPPER, { intent: intentName, entities: entities });
                        }
                        else {
                            session.beginDialog(intentName, entities);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    BlisDialog.prototype.CallLuisCallback = function (text, predictedEntities, memory, allEntities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var memoryManager, scoreInput;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        memoryManager = new ClientMemoryManager_1.ClientMemoryManager(memory, allEntities);
                        scoreInput = null;
                        if (!BlisDialog.luisCallback) return [3 /*break*/, 2];
                        return [4 /*yield*/, BlisDialog.luisCallback(text, predictedEntities, memoryManager)];
                    case 1:
                        scoreInput = _a.sent();
                        return [3 /*break*/, 4];
                    case 2: return [4 /*yield*/, this.DefaultLuisCallback(text, predictedEntities, memoryManager)];
                    case 3:
                        scoreInput = _a.sent();
                        _a.label = 4;
                    case 4: return [2 /*return*/, scoreInput];
                }
            });
        });
    };
    BlisDialog.prototype.CallBlisCallback = function (scoredAction, memory, allEntities) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var memoryManager, outText;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        memoryManager = new ClientMemoryManager_1.ClientMemoryManager(memory, allEntities);
                        outText = null;
                        if (!BlisDialog.luisCallback) return [3 /*break*/, 2];
                        return [4 /*yield*/, BlisDialog.blisCallback(scoredAction.payload, memoryManager)];
                    case 1:
                        outText = _a.sent();
                        return [3 /*break*/, 4];
                    case 2: return [4 /*yield*/, this.DefaultBlisCallback(scoredAction.payload, memoryManager)];
                    case 3:
                        outText = _a.sent();
                        _a.label = 4;
                    case 4: return [2 /*return*/, outText];
                }
            });
        });
    };
    BlisDialog.prototype.DefaultLuisCallback = function (text, predictedEntities, memoryManager) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _i, predictedEntities_1, predictedEntity, filledEntities, scoreInput;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _i = 0, predictedEntities_1 = predictedEntities;
                        _a.label = 1;
                    case 1:
                        if (!(_i < predictedEntities_1.length)) return [3 /*break*/, 6];
                        predictedEntity = predictedEntities_1[_i];
                        if (!(predictedEntity.metadata && predictedEntity.metadata.positiveId)) return [3 /*break*/, 3];
                        return [4 /*yield*/, memoryManager.blisMemory.BotMemory.ForgetEntity(predictedEntity)];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 5];
                    case 3: return [4 /*yield*/, memoryManager.blisMemory.BotMemory.RememberEntity(predictedEntity)];
                    case 4:
                        _a.sent();
                        _a.label = 5;
                    case 5:
                        _i++;
                        return [3 /*break*/, 1];
                    case 6: return [4 /*yield*/, memoryManager.blisMemory.BotMemory.RememberedIds()];
                    case 7:
                        filledEntities = _a.sent();
                        scoreInput = new blis_models_1.ScoreInput({
                            filledEntities: filledEntities,
                            context: {},
                            maskedActions: []
                        });
                        return [2 /*return*/, scoreInput];
                }
            });
        });
    };
    BlisDialog.prototype.DefaultBlisCallback = function (text, memoryManager) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var outText;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, memoryManager.blisMemory.BotMemory.Substitute(text)];
                    case 1:
                        outText = _a.sent();
                        return [2 /*return*/, outText];
                }
            });
        });
    };
    BlisDialog.prototype.emitError = function (session, err) {
        var m = err.toString();
        err = err instanceof Error ? err : new Error(m);
        session.error(err);
    };
    // Mapping between user defined API names and functions
    BlisDialog.apiCallbacks = {};
    return BlisDialog;
}(builder.Dialog));
exports.BlisDialog = BlisDialog;
